Refactor test helpers

namurphy opened the pull request:
Background from a [previous comment](https://github.com/PlasmaPy/PlasmaPy/pull/605#issuecomment-480386397):

> Many of PlasmaPy’s tests have a simple premise. We check that providing a function or method with a set of positional and keyword arguments leads to the expected outcome. The situation becomes more complex with the realization that the expected outcome may be a string, float, array, or quantity with units. We may expect a specific exception to be raised or a warning to be issued. We may be expecting an infinity or NaN that may or may not have units, and have to take into account that `numpy.inf == numpy.inf` and `numpy.nan == numpy.nan` both evaluate to `False`. We might want to check if values or elements of an array are approximately equal to within a specified tolerance. In all of these cases, we need descriptive error messages that can help new contributors quickly pinpoint what the problem is. Accounting for all of these situations leads to repeated boilerplate code that makes tests harder to write, read, and maintain. This in turn creates an unnecessary barrier to entry for new contributors, especially as unit testing and continuous integration testing are not widely used in much of the plasma physics community.

We currently have a test runner function, [`plasmapy.utils.run_test`](https://github.com/PlasmaPy/PlasmaPy/blob/9ba3f4eb5dbbb1a3d7b25527e0d5eb62c5086edf/plasmapy/utils/pytest_helpers/pytest_helpers.py#L83), that is designed to simplify the process of writing tests by consolidating the functionality needed to compare the actual outcome with the expected outcome.  This function is quite helpful, but it contains a whole bunch of functionality in one place which makes it hard to maintain or extend.  There's also repeated code between `run_test` and `run_test_equivalent_calls`.  

My goal here is to refactor our test runner capabilities to break up the responsibilities in `run_test` into separate classes, each with a single responsibility.  Making it modular like this would make it easier to reuse code, since some of the tests are similar but not quite the same.  As a first step, I am creating classes to represent the expected an actual test outcomes.

Cross-references: #278, #605, #606, #674

pep8speaks commented:
Hello @namurphy! Thanks for submitting your pull request.





* In the file [`plasmapy/utils/pytest_helpers/tests/test_expected.py`](https://github.com/PlasmaPy/PlasmaPy/blob/98b4c6936e7bd04035324a12713fd3d3a46c22c9/plasmapy/utils/pytest_helpers/tests/test_expected.py):

> [Line 14:25](https://github.com/PlasmaPy/PlasmaPy/blob/98b4c6936e7bd04035324a12713fd3d3a46c22c9/plasmapy/utils/pytest_helpers/tests/test_expected.py#L14): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ','
> [Line 31:26](https://github.com/PlasmaPy/PlasmaPy/blob/98b4c6936e7bd04035324a12713fd3d3a46c22c9/plasmapy/utils/pytest_helpers/tests/test_expected.py#L31): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ','
> [Line 51:26](https://github.com/PlasmaPy/PlasmaPy/blob/98b4c6936e7bd04035324a12713fd3d3a46c22c9/plasmapy/utils/pytest_helpers/tests/test_expected.py#L51): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ','
> [Line 93:100](https://github.com/PlasmaPy/PlasmaPy/blob/98b4c6936e7bd04035324a12713fd3d3a46c22c9/plasmapy/utils/pytest_helpers/tests/test_expected.py#L93): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (101 > 99 characters)
> [Line 99:100](https://github.com/PlasmaPy/PlasmaPy/blob/98b4c6936e7bd04035324a12713fd3d3a46c22c9/plasmapy/utils/pytest_helpers/tests/test_expected.py#L99): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (101 > 99 characters)
> [Line 127:1](https://github.com/PlasmaPy/PlasmaPy/blob/98b4c6936e7bd04035324a12713fd3d3a46c22c9/plasmapy/utils/pytest_helpers/tests/test_expected.py#L127): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1

PlasmaPy follows the [PEP8 style guide](https://www.python.org/dev/peps/pep-0008/), which may be checked locally using [pycodestyle](http://pycodestyle.pycqa.org/en/latest/intro.html)

rocco8773 commented:
I suggest that this PR should also try to satisfy issue #746

namurphy commented:
> I suggest that this PR should also try to satisfy issue #746

Agreed!  I was thinking exactly that when I saw #746.

namurphy commented:
After too long and too many distractions, this PR is finally ready for review.   Starting a review in the revised section of `docs/development/testing_guide.rst` should give an idea of the motivation and reasoning.  The highest-level functions are in `runners.py`.  

The changes I made are:

 - I created the `plasmapy.tests.helper` module to host the test helper functionality.  The pytest dependency will therefore be isolated to `plasmapy.tests`.  We therefore would not need to require pytest as a hard dependency if we don't automatically import `plasmapy.tests` (related to #737 and #738).
 - I moved functions to assist with formatting error messages to `plasmapy.utils.formatting` because they have more general uses and do not depend on pytest.  Some changes got propagated through to `plasmapy.atomic` and elsewhere.
 - I created classes in `plasmapy.tests.helper.inputs` to take care of storing/validating the function or class attribute/method and the positional and keyword arguments.  The test is run using the `call` method for each of these classes.
 - I created the `ExpectedTestOutcome` class to store the expected test outcome and allow access to its different properties (`expected.py`).
 - I created the `ActualTestOutcome` class to store the true outcome of the test and allow access to its different properties (`actual.py`).
 - I created the `CompareActualExpected` class to compare instances of `ActualTestOutcome` and `ExpectedTestOutcome`, which sometimes uses the `CompareValues` class to compare values (rather than warnings or exceptions).  
 - Exceptions for tests are stored in `exceptions.py` with names that are meant to be descriptive and give users/developers a good idea of what went wrong.  (Suggestions on better names are welcome.)
 - I removed mutable default arguments in the original pytest helper functions (#752) even though these are likely to be removed soon.  

The things left to do are adding examples to docstrings (especially `runners.py`), update the changelog, look through the code coverage results to see which tests are missing, and hopefully make sure that the tracebacks/error messages are not too cluttered.

I tried to follow [my own advice on writing clean code](https://zenodo.org/record/3491142) by using descriptive names, writing short functions/methods that do exactly one thing, having the code read like a top-down narrative, and writing classes that each have a single responsibility.  At the very least, it's cleaner than other code I've written!

In retrospect, I should have broken this up into multiple smaller PRs, like one for each of the main classes.  Unfortunately, it took me a while to figure out how to fit the different pieces together.

Closes #746

StanczakDominik commented:
Wow, you even added a reviewer's manual, that's awesome :D

I think it's fine to do huge PRs for complex multicomponent pieces of code that need to work together, when you really need to start tinkering with it before you see the optimal way to achieve a goal, as long as the resultant code is clean and well described. Seems like that's going nicely!

I should be able to find some time to get started on reviewing this after Thursday, and I think I should be able to finish by next Tuesday.

StanczakDominik commented:
I started looking through the new code and to be honest I'm having issues grasping the scope of it.

Do you think you could try using the new functionality to simplify some of our existing tests, say, somewhere in particles (not all of them!) and maybe somewhere in the formulary? I think that'd make this much more understandable.

rocco8773 commented on plasmapy/tests/__init__.py:
```diff
@@ -0,0 +1 @@
+from . import helper
```
(*low hanging fruit* 🤣) I feel this sub-package would be more appropriately named as `plasmapy.tests.helpers`. There's more than one helper contained here.

namurphy replied:
Yeah, I was following Astropy's lead since they have `astropy.test.helper`.  I personally would prefer `plasmapy.test.helpers` so we can go with that. :joy_cat:

rocco8773 replied:
Or, we could spend another 2 hrs in the next meeting debating the singular vs plural. 🤣

rocco8773 commented on plasmapy/tests/helper/__init__.py:
```diff
@@ -0,0 +1,29 @@
+    MissingExceptionError,
+    UnexpectedExceptionError,
+    InconsistentTypeError,
+)
```
Try running this through [isort](https://pypi.org/project/isort/) and see how well it organizes the imports.

rocco8773 commented:
I think we should have some ``.. automodapi::`` documentation for the classes and functions defined here.  It would probably help contributors learn how to use the code.  I'm not not sure where to place it at the moment though.

namurphy commented:
This pull request got too large, so I split it up into a bunch of different smaller pull requests: #919, #920, #922, #923, #924, #925, #926, #927, #928, and #929.  The development history that went into those different pull requests is still contained within this one.

Assuming I'm not forgetting anything, the only thing in this pull request that hasn't yet been put in a new pull request is the refactoring of `test_atomic.py` so that it uses the new `plasmapy.tests.helpers.test_runner` instead of the old `plasmapy.utils.pytest_helpers.run_test`.
